test/common/network/utility_test.cc

```diff
@@ -140,6 +218,16 @@ TEST(NetworkUtility, ParseInternetAddress) {
   EXPECT_EQ("[1::2:3]:0", Utility::parseInternetAddressNoThrow("1::2:3")->asString());
   EXPECT_EQ("[a::1]:0", Utility::parseInternetAddressNoThrow("a::1")->asString());
   EXPECT_EQ("[a:b:c:d::]:0", Utility::parseInternetAddressNoThrow("a:b:c:d::")->asString());
+
+  StatusOr<Interface> ifc = getLocalNetworkInterface();
+  if (ifc.ok()) {
+    EXPECT_EQ(
+        absl::StrCat("[fe80::1]:0"),
```

Comment:
This means the address string won't output the scope. Two addresses with different scope will get the same string output. And `Utility::parseInternetAddressNoThrow` is used for the listener to parse the listening address also.

So a question about this, is there any case people want to use this kind of address in the listener? if yes, we have some code in the listener using the address string as the hash map key, for example:
https://github.com/envoyproxy/envoy/blob/60e0a125d84488e5aaf1e60d1770a672076c5e11/source/server/connection_handler_impl.h#L148, that won't work if different address generates same string.

Another question is, does the compare two addresses work for this kind of address?
https://github.com/envoyproxy/envoy/blob/60e0a125d84488e5aaf1e60d1770a672076c5e11/source/common/network/address_impl.cc#L294, maybe some code depends on the address comparing.

Reply:
for q1 I would say in theory yes, if you want to listen on the link local you will need the scope to be properly parse and set on the sockaddr_in6. Regarding the change to the way the string is formed I think it's more tricky. I empirically experienced that there is not a string format that is widely accepted. Example: the kernel likes the scope id as a uint32. The grpc server builder was throwing me an error if I was passing <ipv6>%<scope id> and wanted <ipv6>%<interface name>. 
Because of the above I'm unsure on what is the best path forward, maybe can be to leave the status quo and address it when there is a clear use case?

for q2 I introduced the scope to the Ipv6 base class, updated the `operator==` and added a unit test.

Reply:
> for q1 I would say in theory yes, if you want to listen on the link local you will need the scope to be properly parse and set on the sockaddr_in6. Regarding the change to the way the string is formed I think it's more tricky. I empirically experienced that there is not a string format that is widely accepted. Example: the kernel likes the scope id as a uint32. The grpc server builder was throwing me an error if I was passing % and wanted %. Because of the above I'm unsure on what is the best path forward, maybe can be to leave the status quo and address it when there is a clear use case?
> 

I'm not sure whether we can leave it as, since it will be undefined behavior when people use it. I leave this to the maintainer to make the decision. @htuch @adisuissa 

There are other options here:
* Not using address string as the key. This is todo I left when I added multi-addresses listener, I could help this. But yea, need the maintainer to decide whether we can move this PR forward first.
* Just use '%' in the string output. Another benefit is we have correct log output. But this need to ensure we didn't pass the address string to other lib which doesn't support scope address as you said.

> for q2 I introduced the scope to the Ipv6 base class, updated the `operator==` and added a unit test.

Thanks for the update!

Reply:
Yea, I will take a look at hashing IP address, that is my todo list. And since you already update the code to output the scope in the address string, then we can move forward without waiting on hashing IP address.